onnxruntime/core/mlas/lib/quantize.cpp

```diff
@@ -422,3 +684,160 @@ Return Value:
+    const int8_t* InputB,
+    float ScaleB,
+    int8_t ZeroPointB,
+    float ScaleC,
+    int8_t ZeroPointC,
+    int8_t* OutputC,
+    size_t LengthA,
+    size_t LengthB
+    )
+{
+    MlasQLinearAddKernel<int8_t>(
+        InputA, ScaleA, ZeroPointA,
+        InputB, ScaleB, ZeroPointB,
+        ScaleC, ZeroPointC, OutputC,
+        LengthA, LengthB
+    );
+}
+
+void
+MLASCALL
+MlasQLinearAddU8Kernel(
+    const uint8_t* InputA,
+    float ScaleA,
+    uint8_t ZeroPointA,
+    const uint8_t* InputB,
+    float ScaleB,
+    uint8_t ZeroPointB,
+    float ScaleC,
+    uint8_t ZeroPointC,
+    uint8_t* OutputC,
+    size_t LengthA,
+    size_t LengthB
+    )
+{
+    MlasQLinearAddKernel<uint8_t>(
+        InputA, ScaleA, ZeroPointA,
+        InputB, ScaleB, ZeroPointB,
+        ScaleC, ZeroPointC, OutputC,
+        LengthA, LengthB
+    );
+}
+
+
+template <typename T>
+class MlasQLinearAddPlatformKernel
+{
+public:
+    typedef void (MLASCALL KernelFunction)(
+        const T* InputA,
+        float ScaleA,
+        T ZeroPointA,
+        const T* InputB,
+        float ScaleB,
+        T ZeroPointB,
+        float ScaleC,
+        T ZeroPointC,
+        T* OutputC,
+        size_t LengthA,
+        size_t LengthB
+        );
+
+    typedef KernelFunction* PKernelFunction;
+
+    static PKernelFunction GetKernel();
+};
+
+template <>
+MlasQLinearAddPlatformKernel<uint8_t>::PKernelFunction
+MlasQLinearAddPlatformKernel<uint8_t>::GetKernel()
+{
+    return MlasPlatform.QLinearAddU8Kernel;
+}
+
+template <>
+MlasQLinearAddPlatformKernel<int8_t>::PKernelFunction
+MlasQLinearAddPlatformKernel<int8_t>::GetKernel()
+{
+    return MlasPlatform.QLinearAddS8Kernel;
+}
+
+template <typename T>
+void
+MLASCALL
+MlasQLinearAdd(
+    const T* InputA,
+    float ScaleA,
+    T ZeroPointA,
+    const T* InputB,
+    float ScaleB,
+    T ZeroPointB,
+    float ScaleC,
+    T ZeroPointC,
+    T* OutputC,
+    size_t LengthA,
+    size_t LengthB
+    )
+{
+#if defined(MLAS_TARGET_AMD64)
+        auto kernel = MlasQLinearAddPlatformKernel<T>::GetKernel();
+        kernel(
+#else
+        MlasQLinearAddKernel<T>(
+#endif
+            InputA, ScaleA, ZeroPointA, InputB, ScaleB, ZeroPointB, ScaleC, ZeroPointC, OutputC, LengthA, LengthB);
+}
+
+
+template
+void
+MLASCALL
+MlasQLinearAdd<int8_t>(
+    const int8_t* InputA,
+    float ScaleA,
+    int8_t ZeroPointA,
+    const int8_t* InputB,
+    float ScaleB,
+    int8_t ZeroPointB,
+    float ScaleC,
+    int8_t ZeroPointC,
+    int8_t* OutputC,
+    size_t LengthA,
+    size_t LengthB
+    );
+
+template
+void
+MLASCALL
+MlasQLinearAdd<uint8_t>(
+    const uint8_t* InputA,
+    float ScaleA,
+    uint8_t ZeroPointA,
+    const uint8_t* InputB,
+    float ScaleB,
+    uint8_t ZeroPointB,
+    float ScaleC,
+    uint8_t ZeroPointC,
+    uint8_t* OutputC,
+    size_t LengthA,
+    size_t LengthB
+    );
+
+MLAS_INTERNAL_DATA  MLAS_DECLSPEC_ALIGN(const uint8_t MLasPackBytesMM256VpshufbControl[32], 32) = {
```

Comment:
put the data under #ifdefs for MLAS_TARGET_AMD64?

Reply:
Good idea!